### PR TITLE
Insert Metrics Tracker in the toolchain

### PR DESCRIPTION
**Pull request by Tomcli**
* Add tracker to tracker metrics in the toolchain
* Add option to disable tracker
* Add environment variables to capture tracker option and chatbot name without timestamp.
* Add privacy notice

### CODE REVIEW

**Comment by Tomcli** [comment]
I think my branch doesn't have the cf login credential to pass the build.

File: .bluemix/pipeline.yml [hunk 1]
```diff
@@ -74,6 +80,31 @@ stages:
         echo "https://www.ibmwatsonconversation.com"
         exit 1
       fi
+      
+      # Tracking Metrics
+      if [ "yes" != "$(echo "$DISABLE_METRIC" | awk '{print tolower($0)}')" ]; then
+      space_id=$(cf space ${CF_SPACE} --guid)
+      service_id=$(cf service Bot\ Asset\ Exchange\ Workspaces --guid)
```
Comment (stevemar) [inline .bluemix/pipeline.yml]: nice, didn't know the `--guid` option exists! 

`cf service "Bot Asset Exchange Workspaces" --guid` also works if you want to avoid nasty backslahes

File: README.md [hunk 2]
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
 3. Workspace URL, for example: ``chatbotWorkspaceURL=chatbotWorkspaceURL=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer%2Fraw%2Fmaster%2Fdata%2Fworkspace.json``
 
 Click [this link](https://console.bluemix.net/devops/setup/deploy?repository=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer&chatbotName=Chatbot&chatbotWorkspaceURL=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer%2Fraw%2Fmaster%2Fdata%2Fworkspace.json) to try it out.
+
+## Privacy Notice
+
+Sample DevOps pipeline that include this package may be configured to track deployments to [IBM Bluemix](https://www.bluemix.net/) and other platforms. The following information is sent to a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) service on each deployment:
+
+* Space ID (`space_id`)
```
Comment (stevemar) [inline README.md]: remove "(`space_id`)" here, the reader won't know what that means

File: README.md [hunk 3]
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
 3. Workspace URL, for example: ``chatbotWorkspaceURL=chatbotWorkspaceURL=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer%2Fraw%2Fmaster%2Fdata%2Fworkspace.json``
 
 Click [this link](https://console.bluemix.net/devops/setup/deploy?repository=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer&chatbotName=Chatbot&chatbotWorkspaceURL=https%3A%2F%2Fgithub.com%2FIBM%2Fchatbot-deployer%2Fraw%2Fmaster%2Fdata%2Fworkspace.json) to try it out.
+
+## Privacy Notice
+
+Sample DevOps pipeline that include this package may be configured to track deployments to [IBM Bluemix](https://www.bluemix.net/) and other platforms. The following information is sent to a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) service on each deployment:
```
Comment (stevemar) [inline README.md]: I'd rephrase this as:

"The DevOps pipeline that is included in this package is configured by default with a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) tool. The following information is sent when a user deploys an asset from the Bot Asset Exchange:"

**Review by stevemar — CHANGES_REQUESTED** [review]
a few minor comments, but i need some time to actually try this out, too.

**Review by stevemar — COMMENTED** [review]
sorry i haven't had the time to try this out yet @Tomcli

**Comment by stevemar** [comment]
@Tomcli how did you test this out? Specifically, the UI changes? Was it adapted from somewhere else? Ideally I want to try this out but I'm not sure if we made the flow smart enough to use branches

**Comment by Tomcli** [comment]
All the UI changes are in the deploy.json file. I added a new section that allows users to disable the tracker. You can create a new branch with my changes and try it out with any chatbot by adding `&branch=yourBranchName` at the end of the toolchain link. This way you should able to test the new UI and pipeline code in that branch.

File: .bluemix/deploy.json [hunk 4]
```diff
@@ -32,6 +32,12 @@
             },
             "type": "string",
             "pattern": "\\S"
+        },
+        "metrics-tracker": {
+            "description": {
+                "$ref": "#/messages/deploy.metricsDescription"
+            },
+            "type": "string"
```
Comment (stevemar) [inline .bluemix/deploy.json]: it would be great if this were a dropdown menu like the screenshot below. I looked at the docs and couldn't find the right option though -- the screenshot was from the watson conversation pipeline config

screen shot 2017-11-07 at 12 00 34 am

Comment (stevemar) [reply .bluemix/deploy.json]: maybe type: select? i'm not sure

File: .bluemix/pipeline.yml [hunk 5]
```diff
@@ -74,6 +80,31 @@ stages:
         echo "https://www.ibmwatsonconversation.com"
         exit 1
       fi
+      
+      # Tracking Metrics
+      if [ "yes" != "$(echo "$DISABLE_METRIC" | awk '{print tolower($0)}')" ]; then
```
Comment (stevemar) [inline .bluemix/pipeline.yml]: so as it stands now it's enabled by default

File: .bluemix/nls/messages.yml [hunk 6]
```diff
@@ -7,6 +7,7 @@ deploy.description: "sample toolchain"
 deploy.longDescription: "The Delivery Pipeline automates continuous deployment."
 deploy.appDescription: "The name of your Watson app"
 deploy.appName: "App name"
+deploy.metricsDescription : "The metrics tracker is for IBM to track metrics around deployments of sample applications to IBM Bluemix to measure the usefulness of our examples, so that we can continuously improve the content we offer to you. If you want to disable the metrics tracker, please put down YES."
```
Comment (stevemar) [inline .bluemix/nls/messages.yml]: We should state clearly the default behavior. Please change this to:

"The metrics tracker, which is enabled by default, is for IBM to track the number of times our examples are deployed to IBM Cloud. It is strictly used to measure the usefulness of our examples so that we can continuously improve the content we offer to you. If you want to disable the metrics tracker, please put down YES."

File: .bluemix/pipeline.yml (hunk 5 above)
Comment (stevemar) [reply .bluemix/pipeline.yml]: we should echo out some stuff here...

         echo "============================================================="
         echo "Adding deployment data to metrics tracker... "
         echo "============================================================="

and then again at the end if it was successful uploaded. I say this because I'm not seeing anything in the logs past the following:

```
==================================================
Created new workspace: WORKSPACE_ID: 645bdf74-c71a-4693-86ab-a860bb3eed1e
==================================================
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   310    0   310    0     0   2832      0 --:--:-- --:--:-- --:--:--  2844
{
  "name": "watson-banking-chatbot",
  "created": "2017-11-07T05:19:42.239Z",
  "updated": "2017-11-07T05:19:42.239Z",
  "language": "en",
  "metadata": {
    "runtime_version": "2016-09-20"
  },
  "description": "A cognitive banking chatbot",
  "workspace_id": "645bdf74-c71a-4693-86ab-a860bb3eed1e",
  "learning_opt_out": false,
  "status": "Training"
}

Finished: SUCCESS
```

File: README.md [hunk 7]
```diff
@@ -1,3 +1,5 @@
+![Bluemix Deployments](https://metrics-tracker.mybluemix.net/stats/3dafda64b9efbd06391d5f090bd89113/badge.svg)
```
Comment (stevemar) [inline README.md:1]: i don't think we need this, but meh.

File: README.md [hunk 8]
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
...
+
+## Privacy Notice
+
+The DevOps pipeline that is included in this package is configured by default with a [Deployment Tracker](https://github.com/IBM/metrics-collector-service) tool. The following information is sent when a user deploys an asset from the Bot Asset Exchange:
+
+* Space ID
+* Conversation Service GUID
+* Chatbot Name
+
+This data is collected from the Bluemix DevOps pipeline environment variables in IBM Bluemix. This data is used by IBM to track metrics around deployments of sample applications to IBM Bluemix to measure the usefulness of our examples, so that we can continuously improve the content we offer to you. Only deployments of sample applications that include code to ping the Deployment Tracker service will be tracked.
```
Comment (stevemar) [inline README.md]: replace Bluemix with IBM Cloud here, mention that it's enabled by default

File: README.md [hunk 9]
```diff
@@ -34,3 +36,17 @@ Underneath the covers, we hit the DevOps Service ``https://console.bluemix.net/d
...
+
+## Disabling Deployment Tracking
+
+To disable the tracker, Simply put down `yes` in *Disable Metrics Tracker* under the **Delivery Pipeline** section when you create your Bluemix DevOps pipeline.
```
Comment (stevemar) [inline README.md]: i would say "opt out" instead of "disable" here

**Review by stevemar — COMMENTED** [review]
Thanks @Tomcli -- i left a few more suggestions around wording and logging that I'd like cleaned up before this merges.

I also have a UI suggestion, see if you can get a dropdown menu working instead of a text field for the metrics tracker.

I'll go over this with anamita and show her the UI changes.

I tested this by going to: https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter

screen shot 2017-11-06 at 11 58 52 pm

**Comment by Tomcli** [comment]
Thank you for your feedback @stevemart. I updated the toolchain UI to a dropdown and added some logs for the tracker. Please use the below [link](https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter&branch=tracker) to test the new UI and pipeline code because I am merging from my `tracker` branch.
- https://console.bluemix.net/devops/setup/deploy?repository=https%3A//github.com/Tomcli/chatbot-deployer&chatbotName=Simple%20Banking%20Chatbot&chatbotWorkspaceURL=http://developer.ibm.com/code/wp-content/uploads/sites/118/2017/09/banking.json&cm_sp=dw-bluemix-_-code-_-devcenter&branch=tracker

File: .bluemix/pipeline.yml [hunk 10]
```diff
@@ -74,6 +80,40 @@ stages:
...
+      "space_id": "$space_id",
+      "service_id": "$service_id",
+      "date_sent":"$DATE",
+      "bot_name": "$CHATBOT_NAME_RAW",
+      "config": $config
+      }
+      EOF
+      )
+
+      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST http://metrics-tracker.mybluemix.net/api/v1/track)
```
Comment (stevemar) [inline .bluemix/pipeline.yml]: one last comment, but i think the URL should be a defined variable

**Review by stevemar — APPROVED** [review]
one super final comment, just flip the wording so that if the user selects "YES" they are enabling the tracker, and if they select "NO" they are disabling it. 

positive answers, like "YES", should do positive things.

File: .bluemix/pipeline.yml [hunk 11]
```diff
@@ -105,7 +108,7 @@ stages:
       EOF
       )
 
-      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST http://metrics-tracker.mybluemix.net/api/v1/track)
+      callback=$(curl -d "$data" -H "Content-Type: application/json" -X POST $TRACKER_URL)
       if [[ "$callback" == *"true"* ]]; then
```
Comment (stevemar) [inline .bluemix/pipeline.yml]: you probably want to indent this if block

File: .bluemix/pipeline.yml [hunk 12]
```diff
@@ -82,7 +85,7 @@ stages:
       fi
       
       # Tracking Metrics
-      if [ "YES" != "$DISABLE_METRIC" ]; then
+      if [ "NO" != "$DISABLE_METRIC" ]; then
       echo "============================================================="
```
Comment (stevemar) [inline .bluemix/pipeline.yml]: i just realized this block is not indented.

**Review by stevemar — APPROVED** [review]
@Tomcli merging this change, try to clean up the indentation around the if blocks, they're a bit out of whack

File: .bluemix/pipeline.yml (hunk 12 above)
Comment (Tomcli) [reply .bluemix/pipeline.yml]: Sorry I didn't indent this block because the [EOF](https://github.com/Tomcli/chatbot-deployer/blob/tracker/.bluemix/pipeline.yml#L108) portion of the code doesn't work if there are spaces after it. I will try to find if there's another solution.
